src/_P021_Level.ino

```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+          {
+            if (millis2minutes(timePassedSince(timestamp)) >= (ulong)P021_MIN_TIME)
+            {
+              timestamp         = millis();
+              new_control_state = P021_STATE_IDLE;
+            }
+            else
+            {
+              // Take timestamp if extending from stop condition otherwise keep timestamp from moment output was swiched
+              if (P021_getConfigBit(P021_FLAGS, P021_EXTEND_END))
+              {
+                timestamp = millis();
+              }
+              new_control_state = P021_STATE_EXTEND;
+            }
+          }
+          break;
+
+        // Output is active to extend a started cycle untill minimum time exceeded
+        case P021_STATE_EXTEND:
+
+          if (P021_check_on(value, P021_SETPOINT, P021_HYSTERESIS, invert_input, symetric_hyst, remote_state))
+          {
+            new_control_state = P021_STATE_HEATING;
+          }
+          else if (millis2minutes(timePassedSince(timestamp)) >= (ulong)P021_MIN_TIME)
+          {
+            timestamp         = millis();
+            new_control_state = P021_STATE_IDLE;
+          }
+          break;
+
+        // Output was forced active when operation mode switched to temperature control
+        case P021_STATE_FORCE:
+
+          if (P021_check_on(value, P021_SETPOINT, P021_HYSTERESIS, invert_input, symetric_hyst, remote_state))
+          {
+            // Keep timestamp from moment pump was swiched on
+            new_control_state = P021_STATE_HEATING;
+          }
+          else if (millis2minutes(timePassedSince(timestamp)) >= (ulong)P021_FORCE_TIME)
+          {
+            timestamp         = millis();
+            new_control_state = P021_STATE_IDLE;
+          }
+          break;
       }
-      success = true;
+      break; // switch(opmode) case P021_OPMODE_TEMP, P021_OPMODE_REMOTE
+
+    // Unexpected opmode, force to OPMODE_OFF
+    default:
+      P021_OPMODE = P021_OPMODE_OFF;
+      break;
+  } // switch P021_OPMODE
+
+  // Calculate output state from the newly calculated control state
+  switch (new_control_state)
+  {
+    case P021_STATE_IDLE:
+      relay_output = false; // Relay output state
+      break;
+    case P021_STATE_HEATING:
+      relay_output = true;  // Relay output state
+      break;
+    case P021_STATE_EXTEND:
+      relay_output = true;  // Relay output state
+      break;
+    case P021_STATE_FORCE:
+      relay_output = true;  // Relay output state
+      break;
+    default:                // unexpected state, switch pump off
+      relay_output = false; // Relay output state
       break;
-    }
   }
 
-  return success;
+  // Actuate the output pin taking output invert flag into account
+  if (validGpio(P021_GPIO_RELAY))
+  {
+    relay_output ^= P021_getConfigBit(P021_FLAGS, P021_INV_OUTPUT); // Invert when selected
+    digitalWrite(P021_GPIO_RELAY, relay_output ? HIGH : LOW);
+  }
+
+  #ifdef PLUGIN_021_DEBUG
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
```

Comment:
`PLUGIN_021_DEBUG` can only be enabled if `BUILD_NO_DEBUG` is _not_ defined, causing the build to fail...

Reply:
I hope I did it right. Please review carefully.